omniscidb/QueryEngine/Execute.cpp

```diff
@@ -1796,6 +1813,79 @@ hdk::ResultSetTable Executor::finishStreamExecution(
   return result;
 }
 
+std::pair<std::unique_ptr<policy::ExecutionPolicy>, ExecutorDeviceType>
+Executor::getExecutionPolicyForTargets(const RelAlgExecutionUnit& ra_exe_unit,
+                                       const ExecutorDeviceType requested_device_type,
+                                       const std::vector<InputTableInfo>& query_infos,
+                                       size_t& max_groups_buffer_entry_guess) {
+  if (isDevicesRestricted(ra_exe_unit, requested_device_type)) {
+    LOG(DEBUG1) << "Devices Restricted, falling back on CPU";
+    return {std::make_unique<policy::FragmentIDAssignmentExecutionPolicy>(
+                ExecutorDeviceType::CPU),
+            ExecutorDeviceType::CPU};
+  }
+
+  CHECK(!query_infos.empty());
+  if (!max_groups_buffer_entry_guess) {
+    LOG(DEBUG1) << "The query has failed the first execution attempt because of running "
+                   "out of group by slots. Make the conservative choice: allocate "
+                   "fragment size slots and run on the CPU.";
+    max_groups_buffer_entry_guess = compute_buffer_entry_guess(query_infos);
```

Comment:
Did we move the method into the cost model? Would it create a circular dependency again?

Reply:
We put only the end of this method in the `cost model`, but it seemed to me that it was worth bringing it here in case the `cost model` is disabled. the `cost model` itself does not use this code, so there should be no cyclic dependency. or did I misunderstand the comment?